File: webapp/src/main/java/com/pfe/webapp/service/stats/ProductionStatsService.java
package com.pfe.webapp.service.stats;

import com.pfe.webapp.dto.stats.*;
import com.pfe.webapp.repository.rapport.RendementRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;

@Service
public class ProductionStatsService {

    @Autowired
    private RendementRepository rendementRepository;

    // ========== COMPLETE SUMMARY ==========

    public ProductionStatsDTO getProductionStats(LocalDate startDate, LocalDate endDate) {
        ProductionStatsDTO stats = new ProductionStatsDTO();

        // KPI Cards
        stats.setTotalProductionRecords(rendementRepository.countTotalProductionRecords());
        stats.setAverageProductivity(rendementRepository.getAverageProductivity());
        stats.setActiveTeamsCount(rendementRepository.countActiveTeams());
        stats.setAverageActivityDuration(rendementRepository.getAverageActivityDuration());

        // Charts with date filter
        stats.setProductionByTeam(rendementRepository.getProductionByTeam(startDate, endDate));
        stats.setProductionByActivity(rendementRepository.getProductionByActivity(startDate, endDate));
        stats.setProductionTrend(rendementRepository.getProductionTrend(startDate, endDate));
        stats.setProductionByMission(rendementRepository.getProductionByMission(startDate, endDate));

        // Top 5
        stats.setTop5Teams(getTop5Teams());
        stats.setTop5Activities(getTop5Activities());

        // Advanced (no date filter needed for these)
        stats.setProductivityByTeam(rendementRepository.getProductivityByTeam());
        stats.setProductivityByActivity(rendementRepository.getProductivityByActivity());

        return stats;
    }

    // ========== KPI METHODS ==========

    public Long getTotalProductionRecords() {
        return rendementRepository.countTotalProductionRecords();
    }

    public Double getAverageProductivity() {
        return rendementRepository.getAverageProductivity();
    }

    public Long getActiveTeamsCount() {
        return rendementRepository.countActiveTeams();
    }

    public Double getAverageActivityDuration() {
        return rendementRepository.getAverageActivityDuration();
    }

    // ========== CHART METHODS ==========

    public List<ProductionByTeamDTO> getProductionByTeam(LocalDate startDate, LocalDate endDate) {
        return rendementRepository.getProductionByTeam(startDate, endDate);
    }

    public List<ProductionByActivityDTO> getProductionByActivity(LocalDate startDate, LocalDate endDate) {
        return rendementRepository.getProductionByActivity(startDate, endDate);
    }

    public List<ProductionTrendDTO> getProductionTrend(LocalDate startDate, LocalDate endDate) {
        return rendementRepository.getProductionTrend(startDate, endDate);
    }

    public List<ProductionByMissionDTO> getProductionByMission(LocalDate startDate, LocalDate endDate) {
        return rendementRepository.getProductionByMission(startDate, endDate);
    }

    // ========== TABLE METHODS ==========

    public List<TopTeamDTO> getTop5Teams() {
        List<TopTeamDTO> teams = rendementRepository.getTop5Teams();
        // Add rank manually
        for (int i = 0; i < teams.size(); i++) {
            teams.get(i).setRank(i + 1);
        }
        return teams;
    }

    public List<TopActivityDTO> getTop5Activities() {
        List<TopActivityDTO> activities = rendementRepository.getTop5Activities();
        // Add rank manually
        for (int i = 0; i < activities.size(); i++) {
            activities.get(i).setRank(i + 1);
        }
        return activities;
    }
    // ========== ADVANCED METHODS ==========

    public List<ProductivityByTeamDTO> getProductivityByTeam() {
        return rendementRepository.getProductivityByTeam();
    }

    public List<ProductivityByActivityDTO> getProductivityByActivity() {
        return rendementRepository.getProductivityByActivity();
    }

    // ========== FILTER BY MISSION ==========

    public List<ProductionByTeamDTO> getProductionByTeamByMission(Long missionId) {
        return rendementRepository.getProductionByTeamByMission(missionId);
    }

    public List<ProductionByActivityDTO> getProductionByActivityByMission(Long missionId) {
        return rendementRepository.getProductionByActivityByMission(missionId);
    }
}

File: ngx-admin-master/src/app/services/stats/production-stats.service.ts
// services/stats/production-stats.service.ts

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface ProductionByTeamDTO {
  teamName: string;
  productionValue: number;
}

export interface ProductionByActivityDTO {
  activityCode: string;
  activityName: string;
  productionValue: number;
}

export interface ProductionTrendDTO {
  month: string;
  productionValue: number;
}

export interface ProductionByMissionDTO {
  missionCode: string;
  productionValue: number;
}

export interface TopTeamDTO {
  rank: number;
  teamName: string;
  productionValue: number;
}

export interface TopActivityDTO {
  rank: number;
  activityCode: string;
  activityName: string;
  productionValue: number;
}

export interface ProductivityByTeamDTO {
  teamName: string;
  productionValue: number;
  hoursWorked: number;
  productivity: number;
}

export interface ProductivityByActivityDTO {
  activityCode: string;
  activityName: string;
  productionValue: number;
  hoursWorked: number;
  productivity: number;
}

export interface ProductionStatsDTO {
  totalProductionRecords: number;
  averageProductivity: number;
  activeTeamsCount: number;
  averageActivityDuration: number;
  productionByTeam: ProductionByTeamDTO[];
  productionByActivity: ProductionByActivityDTO[];
  productionTrend: ProductionTrendDTO[];
  productionByMission: ProductionByMissionDTO[];
  top5Teams: TopTeamDTO[];
  top5Activities: TopActivityDTO[];
  productivityByTeam: ProductivityByTeamDTO[];
  productivityByActivity: ProductivityByActivityDTO[];
}

@Injectable({
  providedIn: 'root'
})
export class ProductionStatsService {
  private apiUrl = 'http://localhost:8080/api/stats/production';

  constructor(private http: HttpClient) {}

  getProductionStats(startDate: string, endDate: string): Observable<ProductionStatsDTO> {
    return this.http.get<ProductionStatsDTO>(
      `${this.apiUrl}/summary?startDate=${startDate}&endDate=${endDate}`
    );
  }

  getProductionByTeam(startDate: string, endDate: string): Observable<ProductionByTeamDTO[]> {
    return this.http.get<ProductionByTeamDTO[]>(
      `${this.apiUrl}/by-team?startDate=${startDate}&endDate=${endDate}`
    );
  }

  getProductionByActivity(startDate: string, endDate: string): Observable<ProductionByActivityDTO[]> {
    return this.http.get<ProductionByActivityDTO[]>(
      `${this.apiUrl}/by-activity?startDate=${startDate}&endDate=${endDate}`
    );
  }

  getProductionTrend(startDate: string, endDate: string): Observable<ProductionTrendDTO[]> {
    return this.http.get<ProductionTrendDTO[]>(
      `${this.apiUrl}/trend?startDate=${startDate}&endDate=${endDate}`
    );
  }

  getProductionByMission(startDate: string, endDate: string): Observable<ProductionByMissionDTO[]> {
    return this.http.get<ProductionByMissionDTO[]>(
      `${this.apiUrl}/by-mission?startDate=${startDate}&endDate=${endDate}`
    );
  }

  getTop5Teams(): Observable<TopTeamDTO[]> {
    return this.http.get<TopTeamDTO[]>(`${this.apiUrl}/top5/teams`);
  }

  getTop5Activities(): Observable<TopActivityDTO[]> {
    return this.http.get<TopActivityDTO[]>(`${this.apiUrl}/top5/activities`);
  }

  getProductivityByTeam(): Observable<ProductivityByTeamDTO[]> {
    return this.http.get<ProductivityByTeamDTO[]>(`${this.apiUrl}/productivity/teams`);
  }

  getProductivityByActivity(): Observable<ProductivityByActivityDTO[]> {
    return this.http.get<ProductivityByActivityDTO[]>(`${this.apiUrl}/productivity/activities`);
  }
}